File: frontend/types/task.ts
export type TaskFilter = "today" | "upcoming" | "all" | "done";

export type TaskSummary = {
  id: number;
  title: string;
  description?: string | null;
  status: "TODO" | "IN_PROGRESS" | "DONE";
  dueAt?: string | null;
  completedAt?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type TaskUser = {
  id: number;
  name: string;
  email: string;
};

export type TaskDetail = {
  id: number;
  title: string;
  description?: string | null;
  status: "TODO" | "IN_PROGRESS" | "DONE";
  dueAt?: string | null;
  completedAt?: string | null;
  createdAt: string;
  updatedAt: string;
  creator: TaskUser;
  assignee?: TaskUser | null;
  sharedUsers: TaskUser[];
};

export type TaskCreateRequest = {
  title: string;
  description?: string;
  dueAt?: string;
};

export type TaskUpdateRequest = {
  title: string;
  description?: string;
  dueAt?: string | null;
  assigneeId?: number | null;
};

export type TaskStatusUpdateRequest = {
  status: "TODO" | "IN_PROGRESS" | "DONE";
};

export type TaskShareCreateRequest = {
  userId: number;
};

export type SharedTaskGroups = {
  assignedTasks: TaskSummary[];
  sharedTasks: TaskSummary[];
};

File: src/main/java/com/example/teamflow/task/entity/Task.java
package com.example.teamflow.task.entity;

import com.example.teamflow.share.entity.TaskShare;
import com.example.teamflow.user.entity.User;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

import java.time.LocalDateTime;
import java.util.LinkedHashSet;
import java.util.Set;

@Entity
@Table(name = "tasks")
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private TaskStatus status;

    @Column
    private LocalDateTime dueAt;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "creator_id", nullable = false)
    private User creator;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assignee_id")
    private User assignee;

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<TaskShare> shares = new LinkedHashSet<>();

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @Column
    private LocalDateTime completedAt;

    protected Task() {
    }

    public Task(String title, String description, LocalDateTime dueAt, User creator) {
        this.title = title;
        this.description = description;
        this.dueAt = dueAt;
        this.creator = creator;
        this.status = TaskStatus.TODO;
    }

    @PrePersist
    void prePersist() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
        if (status == null) {
            status = TaskStatus.TODO;
        }
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public Long getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public TaskStatus getStatus() {
        return status;
    }

    public LocalDateTime getDueAt() {
        return dueAt;
    }

    public User getCreator() {
        return creator;
    }

    public User getAssignee() {
        return assignee;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public LocalDateTime getCompletedAt() {
        return completedAt;
    }

    public Set<TaskShare> getShares() {
        return shares;
    }

    public void update(String title, String description, LocalDateTime dueAt, User assignee) {
        this.title = title;
        this.description = description;
        this.dueAt = dueAt;
        this.assignee = assignee;

        if (assignee != null) {
            shares.removeIf(share -> share.getUser().getId().equals(assignee.getId()));
        }
    }

    public void changeStatus(TaskStatus status) {
        this.status = status;
        if (status == TaskStatus.DONE) {
            this.completedAt = LocalDateTime.now();
            return;
        }

        this.completedAt = null;
    }

    public void addShare(User user) {
        shares.add(new TaskShare(this, user));
    }

    public void removeShare(User user) {
        shares.removeIf(share -> share.getUser().getId().equals(user.getId()));
    }
}
